### PR TITLE
#52 Refactor ConsoleDownloadManager to use maven logger instead of sout

### PR DESCRIPTION
**Pull request by longtimeago**
@mickaelistria Hope, you will like it :)

### CODE REVIEW

**Comment by mickaelistria** [comment]
The only issue I see with this change is that some people may have consumed the ConsoleDownloadMonitor class and this change would be interpreted as an API change. Se we should at least increase the version to 1.3.0-SNAPSHOT. What do you think about it?

**Comment by longtimeago** [comment]
@mickaelistria Well, IMHO, `ConsoleDownloadMonitor` is completely internal class, which nobody should depend on. It's a subject to be changed even in patch version.
But I'm still quite new to open-source, so, maybe I'm missing something ?

**Comment by mickaelistria** [comment]
For such OSS projects, the default rule is that anything that is public and not marked as internal explicitly (via package name, annotations or documentation...) is a public API and should be handled as such, with care about versioning.
The rule of semantic versioning, which is recommended for Maven projects, is that any change that isn't backward compatible should increase the major version number (that would be moving to 2.0.0-SNAPSHOT).

For this use-case, maybe you can simply clone the class and apply the improvement suggested on the clone, make the consumer use the new class, and annotate the initial/current one with @Deprecated and a comment to explain that the new one should be used.
If we do that, then there is no API issue and we can stay on a 1.2.2-SNAPSHOT.

**Comment by longtimeago** [comment]
@mickaelistria Thank you for the explanation!
All that you said is 100% true for projects, where consumers are actually using the code by embedding it into their projects as libraries.
In or case all customers are using a plugin as a monolith. The only thing with backward compatibility we should care about is API part. 
And API in our case consists of:
- plugin maven goals;
- plugin configuration parameters

Neither of them we don't change here.
Thus, your points are still weird for me. Am I still missing something important?

**Comment by mickaelistria** [comment]
Typical usage is a maven plugin as a monolith, however there is nothing that has prevented from consuming it as a regular library, so it should respect the rule of a regular library.
That said, my suggestion above about creating a new class and deprecating old one seems to be simple and good enough to not require a version change.

**Comment by longtimeago** [comment]
@mickaelistria

**Comment by longtimeago** [comment]
@mickaelistria Thank you for the clarification!
But still can't fully agree with you as, IMHO, the rule should not be applied to maven plugins in this way.
Do we really care of such weird persons, who decided to use the plugin in that silly way?
Do you know at least one such a person?

By obeying this rule we are really hard limited in changes - none would like to contribute, as current code base is messy (sorry for the truth). It's a big pain to add new feature, as in this case one have to modify existing code. Leaving dead code is also very dangerous.

IMHO, we could state in readme that only plugin API is a subject of backward compatibility control.
In this case I would participate in reworking the codebase.

P.S. It's not hard to fix current PR in a way you would like to see it.
But next time I will think hard before submitting a PR to the project.

Thank you!

**Comment by mickaelistria** [comment]
> IMHO, the rule should not be applied to maven plugins in this way.

This good practice is true for any Java project.

Note that I'm not against such a change, I just initially highlighted that this change has to come with at least a minor version bump, to highlight that there are some changes to consider for adopters. Maybe it would be fine as 1st iteration to create a com.googlecode.mavendownloadplugin.internal package and to move all classes in it and release it as 1.3.0 or 2.0.0, so it would explicit that nothing is to be considered as API at the moment.

**Comment by longtimeago** [comment]
@mickaelistria Well, I'm not against good practices :) 
But, usually, good practices are not applicable to all possible cases.

For example, it's a good practice to create immutable objects. There are lots of benefits come it. 
But this doesn't mean that all objects must be immutable :)

> I don't see have the impression that the project suffers that much from this pseudo-constraint

Really? :) v 1.2.1 was published on Nov 20, 2014.
Maybe, I'm just used to a rapid development ... 

Sorry, for such criticism. I understand your position. You are developing this project for a while and now I'm trying to convince you to break the rules you are used to.

To summarize, let's finish up this discussion.
I'll submit a separate PR according to you proposition:
 
> Maybe it would be fine as 1st iteration to create a com.googlecode.mavendownloadplugin.internal package and to move all classes in it and release it as 1.3.0 or 2.0.0, so it would explicit that nothing is to be considered as API at the moment.

Then I'll rework current PR.

Thank you for your time!

**Comment by mickaelistria** [comment]
@longtimeago should we wait for this to be fixed to release 1.3.0 ?

**Comment by longtimeago** [comment]
@mickaelistria noup, it's not important - no business value.
I'll refresh the PR according to the new package structure after release 1.3.0

**Comment by longtimeago** [comment]
@mickaelistria merged from master and squashed. Please, check.
Thanks!

**Comment by mickaelistria** [comment]
Merged, thanks.
